ghga_validator/plugins/ref_validation.py

```diff
@@ -0,0 +1,249 @@
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Plugin for LinkML JSON Validator used for validating the non inline references"""
+
+from typing import Any, Dict, List, Optional, Union
+
+from linkml_runtime.utils.schemaview import ClassDefinition, SchemaView, SlotDefinition
+from linkml_validator.models import ValidationMessage, ValidationResult
+from linkml_validator.plugins.base import BasePlugin
+from linkml_validator.utils import camelcase_to_sentencecase, snakecase_to_sentencecase
+
+# pylint: disable=too-many-arguments,too-many-locals
+
+
+class RefValidationPlugin(BasePlugin):
+    """
+    Plugin to check whether the values in non inline reference fields point
+    to existing objects.
+
+    Args:
+        schema: Path or URL to schema YAML
+        kwargs: Additional arguments that are used to instantiate the plugin
+
+    """
+
+    NAME = "RefValidationPlugin"
+
+    def __init__(self, schema: str) -> None:
+        super().__init__(schema)
+        self.schemaview = SchemaView(schema)
+
+    def process(self, obj: Dict, **kwargs) -> ValidationResult:
+        """
+        Perform validation on an object.
+
+        Args:
+            obj: The object to validate
+            kwargs: Additional arguments that are used for processing
+
+        Returns:
+            ValidationResult: A validation result that describes the outcome of validation
+
+        """
+        if "target_class" not in kwargs:
+            raise TypeError("'target_class' argument is required")
+        target_class = kwargs["target_class"]
+        messages = []
+        valid = True
+
+        messages = self.validate_json(obj, target_class, obj, target_class, "")
+        if len(messages) > 0:
+            valid = False
+
+        result = ValidationResult(
+            plugin_name=self.NAME, valid=valid, validation_messages=messages
+        )
+        return result
+
+    def get_id_list(
+        self, obj: Dict, root: str, class_name: str, id_slot_name: str
+    ) -> List[str]:
+        """
+        Get the list of all identifiers for the inlined objects
+
+        Args:
+            obj: The object to validate
+            root: The root class name
+            class_name: The name of the class used in range
+            id_slot_name: The identifier slot in the range class
+
+        Returns:
+            List[str]: A list of all identifiers for the inlined objects
+
+        """
```

Comment:
This function only works for top-level inlined objects. Also, it is called everytime you validate a reference and thus searches naively for every reference. Please refactor this so that the known references are iterated only once at the beginning of the validation and so that it works also for objects inlined anywhere in the data.

Reply:
Refactored the implementation: now it is run only once and check all the levels of the JSON for the inlined objects.